File: aaac-membership-system/tools/bulk_import_by_phone.py
#!/usr/bin/env python3
"""
Bulk-create Firebase Auth users by phone number and assign roles/memberId.

Usage:
  python3 tools/bulk_import_by_phone.py \
    --project=<PROJECT_ID> \
    --serviceAccount=/absolute/path/to/serviceAccount.json \
    --csv=/absolute/path/to/members.csv

CSV format (header required):
  phone,memberId,displayName
Example:
  +17734941942,1,HABTAMU BAKANA
"""

import argparse
import csv
import json
import os
import sys

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--project', required=True)
    ap.add_argument('--serviceAccount', required=True)
    ap.add_argument('--csv', required=True)
    args = ap.parse_args()

    # Import Firebase Admin SDK
    try:
        import firebase_admin
        from firebase_admin import credentials, auth, firestore
    except ImportError:
        print('Error: firebase-admin not installed. Run: pip install firebase-admin')
        sys.exit(1)

    # Initialize Firebase Admin
    try:
        cred = credentials.Certificate(args.serviceAccount)
        app = firebase_admin.initialize_app(cred, {'projectId': args.project})
        auth_client = auth
        db = firestore.client()
    except Exception as e:
        print(f'Error initializing Firebase: {e}')
        sys.exit(1)

    created = 0
    linked = 0
    errors = 0

    with open(args.csv, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            phone = row['phone'].strip()
            member_id = row['memberId'].strip()
            display_name = row.get('displayName', '').strip()
            
            if not phone or not member_id:
                continue
                
            try:
                # Try to get existing user by phone
                try:
                    user_record = auth_client.get_user_by_phone_number(phone)
                except auth.UserNotFoundError:
                    user_record = None
                
                # Create user if doesn't exist
                if not user_record:
                    user_record = auth_client.create_user(
                        phone_number=phone,
                        display_name=display_name or None,
                        disabled=False
                    )
                    created += 1
                
                # Set role document
                db.collection('roles').document(user_record.uid).set({
                    'role': 'member',
                    'memberId': member_id
                }, merge=True)
                linked += 1
                print(f'[OK] {phone} → uid={user_record.uid} role=member memberId={member_id}')
                
            except Exception as e:
                errors += 1
                print(f'[ERR] {phone} → {e}')

    print(f'\nDone. created={created}, linked={linked}, errors={errors}')
    sys.exit(1 if errors > 0 else 0)

if __name__ == '__main__':
    main()

File: aaac-membership-system/tools/bulk_import_by_phone.js
#!/usr/bin/env node
/*
Bulk-create Firebase Auth users by phone number and assign roles/memberId.

Usage:
  node tools/bulk_import_by_phone.js \
    --project=<PROJECT_ID> \
    --serviceAccount=/absolute/path/to/serviceAccount.json \
    --csv=/absolute/path/to/members.csv \
    [--region=us-central1]

CSV format (header required):
  phone,memberId,displayName
Example:
  +17734941942,1,HABTAMU BAKANA

Notes:
  - Phone numbers must be in E.164 format (e.g., +15555550123)
  - If a user with the phone already exists, we reuse it
  - Writes roles/{uid} = { role: 'member', memberId }
*/

const fs = require('fs');
const path = require('path');
const readline = require('readline');

function parseArgs() {
  const out = {};
  for (const arg of process.argv.slice(2)) {
    const m = arg.match(/^--([^=]+)=(.*)$/);
    if (m) out[m[1]] = m[2];
  }
  return out;
}

async function main() {
  const args = parseArgs();
  const { project, serviceAccount, csv, region } = args;
  if (!project || !serviceAccount || !csv) {
    console.error('Missing args. Required: --project, --serviceAccount, --csv');
    process.exit(1);
  }

  const saPath = path.resolve(serviceAccount);
  const csvPath = path.resolve(csv);
  if (!fs.existsSync(saPath)) {
    console.error('Service account JSON not found:', saPath);
    process.exit(1);
  }
  if (!fs.existsSync(csvPath)) {
    console.error('CSV not found:', csvPath);
    process.exit(1);
  }

  const admin = require('firebase-admin');
  const sa = require(saPath);
  admin.initializeApp({
    credential: admin.credential.cert(sa),
    projectId: project,
  });
  const auth = admin.auth();
  const db = admin.firestore();
  if (region) db.settings({ }); // placeholder; Firestore auto-region

  let created = 0, linked = 0, errors = 0;

  const rl = readline.createInterface({ input: fs.createReadStream(csvPath), crlfDelay: Infinity });
  let lineNo = 0;
  for await (const line of rl) {
    lineNo += 1;
    if (!line.trim()) continue;
    if (lineNo === 1 && /^\s*phone\s*,\s*memberId/i.test(line)) continue; // header
    const parts = line.split(',');
    if (parts.length < 2) { console.warn('Skip line (need at least phone,memberId):', line); continue; }
    const phone = parts[0].trim();
    const memberId = String(parts[1]).trim();
    const displayName = (parts[2] || '').trim();
    if (!/^\+\d{7,15}$/.test(phone)) { console.warn('Invalid phone (must be E.164):', phone); continue; }
    if (!memberId) { console.warn('Missing memberId for phone:', phone); continue; }

    try {
      let userRecord;
      try {
        userRecord = await auth.getUserByPhoneNumber(phone);
      } catch (_) {
        userRecord = null;
      }

      if (!userRecord) {
        userRecord = await auth.createUser({ phoneNumber: phone, displayName: displayName || undefined, disabled: false });
        created += 1;
      }

      await db.collection('roles').doc(userRecord.uid).set({ role: 'member', memberId }, { merge: true });
      linked += 1;
      console.log(`[OK] ${phone} → uid=${userRecord.uid} role=member memberId=${memberId}`);
    } catch (e) {
      errors += 1;
      console.error(`[ERR] ${phone} →`, e && e.message ? e.message : e);
    }
  }

  console.log(`\nDone. created=${created}, linked=${linked}, errors=${errors}`);
  process.exit(errors ? 1 : 0);
}

main().catch(e => { console.error(e); process.exit(1); });



